### PR TITLE
Use 'pre_wp_unique_post_slug' for unique post name

### LINKED ISSUE #44: `wp_unique_post_slug()` overhead on large sites
Keeping this short and sweet, more as a note / for documentation.

`wp_unique_post_slug()` is called on some action scheduler related wp-cron.php calls when actions are marked complete etc.

On sites with a large number of items in the posts table this adds a substantial overhead  e.g.  

new relic trace

Looks like its bit of a known bottleneck for large sites http://wordpress.stackexchange.com/questions/119738/can-we-have-a-post-without-a-slug

As the actions aren't really public, calculating a slug/name is not really necessary.

Unfortunately there is no easy way to get around it as it is used within `wp_update_post()` e.g. https://github.com/Prospress/action-scheduler/blob/master/classes/ActionScheduler_wpPostStore.php#L563 and the only filters within in are at [the end](https://github.com/WordPress/WordPress/blob/5031ca5c23f68b20fd63fd8370c05e9b0e1af164/wp-includes/post.php#L3675)

An action or filter at the beginning around [here](https://github.com/WordPress/WordPress/blob/5031ca5c23f68b20fd63fd8370c05e9b0e1af164/wp-includes/post.php#L3553) would potentially allow us to hijack and ignore what happens after.

[This trak ticket](https://core.trac.wordpress.org/ticket/21112) may or may not be useful/interesting. Its certainly related.

For more discussion see https://github.com/Prospress/woocommerce-subscriptions/issues/1355

### PR DESCRIPTION
**Pull request by thenbrent**
**tl;dr** - WordPress 5.1 is [introducing the `'pre_wp_unique_post_slug'` filter](https://core.trac.wordpress.org/ticket/21112#comment:27) for plugins to short-circuit the non-performant approach it takes to determining unique post slugs/`post_name` values.

We can take advantage of this filter to fix #44 as an interim perf improvement until #77 is finished.

---

When an action's post status is transitioned to something other than `'draft'`, `'pending'` or `'auto-draft'`, like `'publish'` or `'failed'` or `'trash'`, WordPress will find a unique slug (stored in the `post_name` column of the `wp_posts` table) using the `wp_unique_post_slug()` function.

This is done to ensure URL uniqueness for posts. It's not needed for scheduled action posts, which have no URL.

The approach taken by `wp_unique_post_slug()` is to iterate over existing `post_name` values that match, and increment a numeric suffix until a unique string is found. This makes sense when manually creating a post from the **Edit Post** screen. It becomes a bottleneck when automatically processing thousands of actions, with a database containing thousands of related `post_name` values.

WordPress 5.1 is introducing the `'pre_wp_unique_post_slug'` filter for plugins to address this issue.

We can short-circuit WordPress's `wp_unique_post_slug()` approach using the `'pre_wp_unique_post_slug'` filter. This patch adds `ActionScheduler_wpPostStore::set_unique_post_slug()` to be used as a callback on that filter. It provides a more scalable approach to generating a `post_name`/slug that is probably unique. Because Action Scheduler never actually uses the `post_name` field, or an action's slug, being probably unique is good enough.

For more backstory on this issue, see:
- https://github.com/Prospress/action-scheduler/issues/44 and
- https://core.trac.wordpress.org/ticket/21112

### CODE REVIEW

**Comment by thenbrent** [comment]
@WPprodigy yeah I considered that. Action Scheduler itself doesn't use the `post_name` at all, but I thought it better to at least attempt uniqueness to avoid any possible gotchas with either 3rd partly plugins or our future selves that may assume uniqueness.

**Comment by james-allan** [comment]
This looks good to me.

I've run a few tests locally and everything looks fine. A couple of things I noticed:

We apply our custom post name/slug on all post statuses and so where trashed (cancelled) scheduled actions would have had `__trashed-107` they now have our unique slug. eg http://pic.pros.pr/f71b503302b6.

I think I mentioned the same would happen for pending scheduled actions, however that doesn't appear to be the case. Pending scheduled actions still have empty post names (http://pic.pros.pr/c6c9a96407d0 - two pending scheduled actions created before and after this change). This is because WP won't generate post names for posts with a 'draft', 'pending', or 'auto-draft' status. 

These are things I'm not really concerned about at all but something we mentioned in our chat IRL last week. I'm not sure if it's worth making those type of things backwards compatible. Especially since it is going to be less performant to let WP generate those unique trashed post names.  

Something else we talked about was whether we should check if the slug passed in is `null`. If someone else has overrriden the slug and returned theirs earlier then ours, should we honor their's? This could give folks a way to generate an even more performent one if they'd like to bypass ours.

Other than that, this looks good to me. 👍

**Comment by james-allan** [comment]
Alright with that then this looks good to go. 🚀👍
